Image Based Picking

danieldresser-ie opened the pull request:
This is definitely not done yet, but John thought it could be useful to take a look at the work-in-progress so we can start working on the APIs involved before we finish nailing down all the actual interactions.

I'm trying to think of all the things left to do:
* ImagePickTool needs to communicate when things go wrong via the tool overlay in the lower left
* Various things need to detect when metadata changes, and update dynamically.
* ImagePickTool should probably support dragging a selection rectangle?
* We wanted to support Cryptomatte as well - we should detect if there's a Cryptomatte id manifest, and use that.
* We need to decide on how we actually want to stylize the selection rendering.

There are lots of TODOs in the code. But the basics all seem to be working, and the approach mostly seems pretty reasonable.

-Daniel

danieldresser-ie commented:
I should probably put together a clean demo scene on Monday for you to play with, but the key you need to get this working is just these settings:

```
import Gaffer
import GafferScene
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 5, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 8, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["CustomOptions1"] = GafferScene.CustomOptions( "CustomOptions1" )
parent.addChild( __children["CustomOptions1"] )
__children["CustomOptions1"]["options"].addChild( Gaffer.NameValuePlug( "", Gaffer.StringPlug( "value", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member1", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["CustomOptions1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs7"] = GafferScene.Outputs( "Outputs7" )
parent.addChild( __children["Outputs7"] )
__children["Outputs7"]["outputs"].addChild( Gaffer.ValuePlug( "output2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs7"]["outputs"]["output2"].addChild( Gaffer.StringPlug( "name", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs7"]["outputs"]["output2"].addChild( Gaffer.BoolPlug( "active", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs7"]["outputs"]["output2"].addChild( Gaffer.StringPlug( "fileName", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs7"]["outputs"]["output2"].addChild( Gaffer.StringPlug( "type", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs7"]["outputs"]["output2"].addChild( Gaffer.StringPlug( "data", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs7"]["outputs"]["output2"].addChild( Gaffer.CompoundDataPlug( "parameters", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Outputs7"]["outputs"]["output2"]["parameters"].addChild( Gaffer.NameValuePlug( "layerPerLightGroup", Gaffer.BoolPlug( "value", defaultValue = False, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), "layerPerLightGroup", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs7"]["outputs"]["output2"]["parameters"].addChild( Gaffer.NameValuePlug( "", Gaffer.StringPlug( "value", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member1", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs7"]["outputs"]["output2"]["parameters"].addChild( Gaffer.NameValuePlug( "", Gaffer.StringPlug( "value", defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member2", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs7"]["outputs"]["output2"]["parameters"].addChild( Gaffer.NameValuePlug( "", Gaffer.BoolPlug( "value", defaultValue = False, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member3", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["Outputs7"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["CustomOptions1"]["options"]["member1"]["name"].setValue( 'render:idManifestFilePath' )
__children["CustomOptions1"]["options"]["member1"]["value"].setValue( '${project:rootDirectory}/renders/${script:name}/${renderPass}/id/idManifest.####.exr' )
__children["CustomOptions1"]["__uiPosition"].setValue( imath.V2f( 243.322083, 150.188904 ) )
__children["Outputs7"]["in"].setInput( __children["CustomOptions1"]["out"] )
__children["Outputs7"]["outputs"]["output2"]["name"].setValue( 'Batch/Arnold/Id' )
__children["Outputs7"]["outputs"]["output2"]["fileName"].setValue( '${project:rootDirectory}/renders/${script:name}/${renderPass}/id/id.####.exr' )
__children["Outputs7"]["outputs"]["output2"]["type"].setValue( 'exr' )
__children["Outputs7"]["outputs"]["output2"]["data"].setValue( 'int id' )
__children["Outputs7"]["outputs"]["output2"]["parameters"]["member1"]["name"].setValue( 'filter' )
__children["Outputs7"]["outputs"]["output2"]["parameters"]["member1"]["value"].setValue( 'closest' )
__children["Outputs7"]["outputs"]["output2"]["parameters"]["member2"]["name"].setValue( 'layerName' )
__children["Outputs7"]["outputs"]["output2"]["parameters"]["member2"]["value"].setValue( 'id' )
__children["Outputs7"]["outputs"]["output2"]["parameters"]["member3"]["name"].setValue( 'header:gaffer:loadAsUint' )
__children["Outputs7"]["outputs"]["output2"]["parameters"]["member3"]["value"].setValue( True )
__children["Outputs7"]["__uiPosition"].setValue( imath.V2f( 243.322083, 142.024841 ) )

del __children```

johnhaddon commented on src/GafferSceneUI/ImagePickTool.cpp:
```diff
@@ -0,0 +1,440 @@
+		throw IECore::Exception( "No manifest found" );
+	}
+
+	const Imf::Header header = exrInput.header( idManifestPart );
+	int identifier = header.typedAttribute<Imf::IntAttribute>( "gaffer:idManifestIdentifier" ).value();
+
+	if( identifier != m_sideCarManifestIdentifier )
+	{
+		// TODO - this should probably be some sort of overlay message, not an exception?
+		throw IECore::Exception( "Manifest identifier doesn't match" );
+	}
```
I do still wonder if this identifier check is maybe a little too fussy - do you think it's particularly likely that things will get out of sync?

danieldresser-ie replied:
It's probably not necessary - we only need the id stored in the actual image AOVs to trigger refreshes.

My thought process was basically just: if we have something called a manifestIdentifier, it makes sense for it to go in the manifest file. And then if it's in the manifest file, we might as well check it.

If you don't think it's worth checking here, is it even worth writing it to the manifest file? I guess hypothetically in a scenario where something has gone wrong with this process, someone could manually inspect the headers and observe the ids don't match ... but the chances of anyone knowing about the id to check seems pretty low? Maybe we just don't write it to the manifest file?

To be honest, I've forgotten what your argument was for why we want to call this an id rather than a timeStamp, if it's only function is to trigger reloading the manifest if the AOVs indicate that the manifest has been modified.

johnhaddon replied:
> If you don't think it's worth checking here, is it even worth writing it to the manifest file? 

That's what I was getting at - do we need this at all? @murraystevenson might know better, but I don't think Cryptomatte has an equivalent, and I'm not aware of it causing significant issues (although I have little direct experience with it). 

> I've forgotten what your argument was for why we want to call this an id rather than a timeStamp

I think it was to allow for pipelines where render outputs are written locally and then moved/renamed/post-processed?

danieldresser-ie replied:
> I think it was to allow for pipelines where render outputs are written locally and then moved/renamed/post-processed?

That's an argument for not validating that the timestamp matches the file on disk, same as we're thinking of not validating the id. Maybe it's more weird not to validate it if it's a timestamp?

I'm trying to think of a name that says "use this to trigger updates, but nothing else". It's almost like an "updateCounter" - except that might imply it only increases by 1 when you render, rather than increasing every millisecond?

johnhaddon replied:
I haven't looked at the implementation enough to have a clear idea on this, but it seems like there are two separate concerns :

1. Knowing when to reload the manifest.
2. Validating that the manifest corresponds to the ID render.

I'm not too fussed about 2 - we don't do anything to validate that all AOVs come from the same render (and there are times when you know what you're doing and want to fix and match).

For 1, is it beneficial for the update trigger to be in the image metadata? Would it be possible for the trigger to be the timestamp of the file itself? Or is there no good point at which to check for changes?

danieldresser-ie replied:
To me, it makes a lot more sense in this case if the trigger is coming from the image metadata.

Suppose at some point in time I am happily viewing id.exr, which points to manifest.exr, and everything is working.

If at some future point, manifest.exr changes on disk, that isn't really what should trigger an update ... I've already got a manifest which works with my current ids?

What needs to trigger an update is if you receive a new ids pass - then you need to read a new manifest that matches the new ids?

johnhaddon replied:
We definitely need the metadata to trigger updates - the manifest filename changing would be the simplest reason for that. But perhaps we don't need a special key in the metadata in addition to the filename? What if when the metadata is dirtied, we do two things : check the filename is the same, and if it is, check that the timestamp of the file we loaded already is the same as the timestamp the file has now.

danieldresser-ie replied:
Hmm, I can't think of any reason that wouldn't work, and I do kind of like the idea of removing an extraneous piece of state.

It does feel a bit weird to me ... trying to think why it feels odd. I think it's because usually in Gaffer, I wouldn't expect emitting dirty on a plug where no hashes or values have changed to change the result the user is seeing ( unless someone has messed up hash() or affects() ). But I guess that's not a law, and there's no reason we can't leverage this to pick times to do a stat().

johnhaddon replied:
Yeah, we're not writing a ComputeNode here - I think it's OK for the UI to use slightly different logic.

johnhaddon commented on src/GafferImage/OpenImageIOReader.cpp:
```diff
@@ -1381,14 +1381,33 @@ class File
-					regionRect.min.x, regionRect.max.x, regionRect.min.y, regionRect.max.y,
-					0, 1, 0, spec.nchannels, TypeDesc::FLOAT, &buffer[0]
-				) )
+				if( spec.format == TypeDesc::UINT32 && spec.get_int_attribute( "gaffer:loadAsUint", 0 ) )
```
I'm not super keen on "gaffer:loadAsUint" - seems to offer some public functionality we might not want to actually support that widely. Are there any other options? If we loaded all UINT32 channels this way, do you think there would be any casualties? Is the default OIIO behaviour useful to anyone in this case? Or should we use `gaffer:thisIsAnIDRender` to limit the behaviour to just our ID renders and nothing else?

danieldresser-ie replied:
I was a bit surprised when after suggesting a super specific name here, you decided you liked the proposal I tossed out of doing something very generic like "gaffer:integerMapping" ( supporting "remap", "reinterpret", or in the future "cast" ). I guess being able to handle the "cast" case properly appeals to you? It does raise a naming question though - if we're thinking about long term cleanliness, I don't think it makes sense to have a separate "gaffer:uintMapping" - we'd want the same metadata to affect both. So if it's named "gaffer:integerMapping", should I implement it for int32 as well as uint32?

The part that definitely doesn't need to get implemented now, but it is maybe worth thinking about for how this is presented, is that in the long run, if we want to offer this control, we would probably want to offer these options as overrides on ImageReader ( this would allow you to load your ids in "cast" mode if you have less than 16 million of them and want to interact with them using standard Gaffer compositing tools ).

johnhaddon replied:
> I was a bit surprised when after suggesting a super-specific name here, you decided you liked the proposal I tossed out of doing something very generic like "gaffer:integerMapping"

Yeah, that is hard to explain :)

I definitely still do not like `loadAsUint` though. It sounds like it could apply to any source type, and would convert to a `uint` in Gaffer. I like that `integerMapping` sounds like it will apply to integers, and I like that `reinterpret` sounds just as sketchy as it does.

Regardless of what we call this, how does it interact with your new multi-part EXR support for Arnold outputs? Is there any way we can make this apply only to the ID channel if there are other integer channels in the file?

johnhaddon commented on include/GafferImageUI/ImageGadget.h:
```diff
@@ -167,6 +170,15 @@ class GAFFERIMAGEUI_API ImageGadget : public GafferUI::Gadget
+
+		void setHighlightId( uint32_t id );
+
+		// Copied from GafferSceneUI/Private/OutputBuffer.h
+		// TODO : Currently public so it can be accessed from TileShaderSelectedIds::ScopedBinding
+		class BufferTexture;
```
This feels small enough that I wouldn't mind duplicating it to keep it internal until such a time as IECoreGL provides it.

johnhaddon reviewed (commented):
Thanks Daniel - this seems to be coming together nicely. I particularly like how the single render option for the location of the manifest file is all that is needed to set this up. Seems like that should go in StandardOptions?

I've limited my review comments to high-level design issues or user-facing behaviours, and haven't put any effort into detailed review. Hopefully that's reasonable given that this is just an initial draft. I've made comments in different places that all point towards PathIdMap becoming a slightly more fully-featured Manifest class which encapsulates the IO side of things too - I think that allows PathIdMap to encapsulate its implementation better, and also puts all the useful public API in a single place.

I also tested the tools with the outputs from a batch render, and all largely worked as expected. I did manage to get into a weird state where I could click to select things and it selected them in the HierarchyView, but nothing was highlighted in the Viewer. And I also managed to break things by switching from an image with Ids to one without, which lead to this error :

```
OpenGL.error.GLError: GLError(
        err = 1282,
        description = b'invalid operation',
        baseOperation = glPopClientAttrib,
        cArguments = ()
)
```

I suspect both are down to not updating something at the right moment based on changes to the image - seems like you already had that on your "up next" list anyway.

It also seems that we get hover highlighting on the background, where I wouldn't expect there to be a valid ID - I found this quite distracting. And there's always an outline between the objects with an ID and the background which is also distracting. I think it will be hard to judge the overall effectiveness of the drawing scheme without those being fixed first.

I didn't test with an InteractiveRender, but from seeing that the controller is now exposed publicly, I assume the tool works with that by getting the manifest from the controller directly? If so, do we also write the manifest file when we stop the render, so the catalogue images continue to work?

Hopefully those comments are useful...

Cheers...
John

danieldresser-ie commented on resources/graphics.py:
```diff
@@ -206,6 +206,7 @@
 				'gafferSceneUILightTool',
 				'gafferSceneUILightPositionTool',
 				'gafferSceneUIVisualiserTool',
+				'gafferSceneUIImagePickTool',
```
Do you think we should have a cleaner way of overriding the icon for a tool, or is it OK that I have to do this:
```
    plugs = {
        "active" : [
            "boolPlugValueWidget:image", "gafferSceneUISelectionTool.png"
        ],
    }
```

( I'm assuming I shouldn't be duplicating the SVG just to reuse the icon? )

johnhaddon replied:
I think what you've done there is totally reasonable. The Tool base class provides a default metadata value, and you're overriding it - that's how metadata is intended to work.

danieldresser-ie commented on python/GafferSceneUI/ImagePickToolUI.py:
```diff
@@ -0,0 +1,167 @@
+	"""
+	""",
+
+	"viewer:shortCut", "C",
```
Seems like a good idea, but Q is currently toggling image comparison in an image context. Should we use a different hotkey for that? Is @murraystevenson the current arbiter of hotkeys?

danieldresser-ie commented on include/GafferScene/PathIdMap.h:
```diff
@@ -0,0 +1,91 @@
+	public :
+		PathIdMap( bool threadsafe );
+
+		uint32_t mapPath( const ScenePlug::ScenePath &path );
+		void insertPath( const ScenePlug::ScenePath &path, uint32_t id );
+
+		uint32_t idForPath( const ScenePlug::ScenePath &path ) const;
+		std::optional<ScenePlug::ScenePath> pathForId( uint32_t id ) const;
```
I'm happy to rename to RenderManifest, and move the code to read/write to EXR here - I think that will naturally address some of the issues ( including getting rid of insertPath ).

As for some of the API changes: I really didn't like the old naming - idForPath and pathForID sound like they are parallel, but actually one of them is const, and just return a match if it's found, whereas the other requires write access because it's responsible for adding new paths to the mapping based a counter ... it felt a lot more appropriate to give the insertion function a different name? If we do want to keep the name idForPath for the function that inserts new entries, we probably need a new name for a const function that just does a lookup ( I think it is worthwhile to be able to pass around a const RenderManifest to ImagePickTool, and still do lookups ).

> What's the reason for the threadsafe argument?

It felt silly to be constantly acquiring and releasing the mutex while reading in a potentially large EXR manifest - if we're moving that inside RenderManifest, that solves one part of it, since that function will no longer be reliant on the public API. Maybe it would be nice to add plural idsForPaths and pathsForIds functions to RenderManifest ( since both RenderController and ImagePickTool need them ). At that point, there wouldn't be any downside to always taking the mutex, even in cases where we know there can't possibly be contention. I haven't actually benchmarked this, so maybe it wouldn't make much difference anyway, but it felt wrong to be acquiring a mutex in cases where we aren't using it ... maybe idsForPaths and pathsForIds are good ideas anyway, and then this won't matter.

johnhaddon replied:
> idForPath and pathForID sound like they are parallel, but actually one of them is const, and just return a match if it's found, whereas the other requires write access because it's responsible for adding new paths to the mapping based a counter ... it felt a lot more appropriate to give the insertion function a different name?

OK, I buy that. Definitely worth addressing now that this is going to be public API.

May I suggest `acquireId()` for the non-const function? We use that terminology almost everywhere to mean "get me a thing, either a new one, or one you made earlier, so the thing will be shared with other clients".

> maybe idsForPaths and pathsForIds are good ideas anyway

They sound good to me, especially if it means we can get rid of the threadsafe argument.

danieldresser-ie commented on src/GafferScene/RendererAlgo.cpp:
```diff
@@ -1684,6 +1698,13 @@ struct ObjectOutput : public LocationOutput
 				m_lightLinks->outputLightLinks( scene, attributes(), objectInterface.get() );
 			}
+
+			if( m_idCounter )
+			{
+				int id = m_idCounter->fetch_add( 1 );
+				m_threadIdList->local().push_back( { path, id } );
+				objectInterface->assignID( id );
```
Having every thread hitting a spinlock whenever it tries to write an id seemed pretty inefficient compared to using thread local storage ... we don't even need to build our map data structure here, we're just passing ids to EXR to write out.

If we just use PathIdMap though, including it's id assignment mechanism, it does seem to make things simpler, so I figured I should quickly hack things up to get some sort of measurement of what the actual cost is.

If my quick hacking is accurate, the numbers seem to be working out like this:

* Rendering 1 000 000 spheres with `scene:render:sceneTranslationOnly` - just existing Gaffer and Arnold overhead : 10.2s
* Accumulating all the ids using thread local vectors: 0.1s
* Accumulating all the ids using PathIdMap: 0.4s
* Writing the ids to a compressed exr manifest: 2.5s

Looking at that, if you want to make the call that the accumulation time is a small enough part of the overall time here that it would be worth the simplicity of just using PathIdMap ( or RenderManifest as it's going to be called ), then that's probably not unreasonable.

johnhaddon replied:
Yep, I think I'm willing to make that call.

The other thing here is that we've been using PathIdMap in RenderController for a very long time, and there we care even more about time to first pixel because it's used for InteractiveRender and the Viewer. So if we think performance is an issue, we should improve the performance for PathIdMap rather than avoid using it.

danieldresser-ie commented:
There were enough odds and ends to finish up that I haven't done the Python binding and tests for RenderManifest, but I think just about everything else is ready for review.

There was a bit of awkwardness with the API for RenderManifest when it comes to Cryptomatte, because the process of finding the manifest from the image metadata is pretty tightly linked to telling whether the manifest has changed? It seems pretty important to be using a timestamp or somereadCryptomatteManifestFromFilePaththing to avoid repeatedly reading the manifest off disk? If I implement the manifest search logic inside RenderManifest, I'd be reimplementing the same logic in ImageSelectionTool to check if it's changed. It doesn't really make sense to put the cache logic in RenderManifest ( especially if the RenderManifest function is static ).

Currently, I've just ended up putting all the search logic in ImageSelectionTool, and exposing two functions in RenderManifest: readCryptomatteManifestFromString and readCryptomatteManifestFromFilePath, which parse a boost property_tree from a file, or a string in memory. Other alternatives: I could do one function that takes an std::istream? Or I could just take an std::string, and not care about the possibility that loading the whole string into memory first might be inefficient when parsing a large file ( as long as we're dealing with Cryptomatte files that seem to choke after 16 thousand entries, it's probably fine if we aren't parsing them efficiently ).

johnhaddon commented:
Thanks Daniel! I realise looking at the ImageGadget changes that we're going to have to target most of this to `main` for inclusion in 1.6 (because of the ABI breaks). I think that's fine though - it's high time we started focusing on 1.6 anyway.

I think there's a fairly natural split where everything up to and including https://github.com/GafferHQ/gaffer/pull/6329/commits/a7a1815fabdf797495571839e579a07fa58cd5e3 (but omitting [f50670a](https://github.com/GafferHQ/gaffer/pull/6329/commits/f50670a536c92cc5da56fedcf754a38f1ed9f792)) makes sense as incremental improvements to 1.5. Could you update this PR to include only those commits please, and then once it's merged we can focus on getting the rest into `main`?

johnhaddon commented:
> Could you update this PR to include only those commits please

Sorry, should have mentioned this in my previous comment - could you please also update Changes.md as appropriate in those commits? Thanks!

danieldresser-ie commented:
I've split out the prerequisites you've requested for 1.5 into #6398.

danieldresser-ie commented:
Retargeted to main, and rebased onto #6398

johnhaddon commented on src/GafferScene/RenderManifest.cpp:
```diff
@@ -0,0 +1,445 @@
+
+	for( Imf::IDManifest::ChannelGroupManifest::ConstIterator i = idManifest.begin(); i != idManifest.end(); ++i )
+	{
+		m_map.insert( PathAndID( ScenePlug::stringToPath( i.text()[0] ), i.id() ) );
```
`acquireId()` assumes that it is the only creator of IDs, and can use `m_map.size()` to generate the next ID. Since we're loading arbitrary IDs here, that assumption is broken.

I guess this kindof brings us back to the question of whether or not `load()` should have been static (and return a const RenderManifest). One possibility would be to make it static, but have a cache behind the scenes. Memory usage of such a cache might be a worry, but we could use `weak_ptr` in that so we only kept things as long as our clients do - since what we're trying to optimise is a client reloading the same data again, that could work?

danieldresser-ie replied:
I don't really like making the cache a global map if we don't need to. Two alternatives:

* Store a bool for whether the RenderManifest has been modified in any way other than acquireID. If this bool is set, acquireID throws. clear() resets it.
* acquireID could take an argument for the id to use if the path isn't found. The existing use cases of acquireID could pass in renderManifest->size() for this argument, since they know that aren't modifying the manifest in any other way.

johnhaddon replied:
> acquireID could take an argument for the id to use if the path isn't found.

That seems to defeat the point of `acquireID()` - to get an ID that everyone can share, without needing to know what it should. What if the ID already exists, and the ID I provide doesn't match it? Just seems fragile.

> Store a bool for whether the RenderManifest has been modified in any way other than acquireID. If this bool is set, acquireID throws. clear() resets it.

This is at least honest about the restriction, and is unlikely to be a problem because we don't have a use case for load-then-modify. But it still seems ugly to have to track additional state to tell us that a part of the API we provide is off limits.

> I don't really like making the cache a global map if we don't need to.

Can you explain why? `static ConstPtr load()` is more natural, and has no API holes like the above do (well, at least until we lose constness in Python). And `weak_ptr` means there is basically no overhead in terms of memory.

johnhaddon commented on src/GafferScene/RenderManifest.cpp:
```diff
@@ -0,0 +1,445 @@
+
+void RenderManifest::writeEXRManifest( const std::filesystem::path &filePath ) const
+{
+	if( !boost::ends_with( filePath.generic_string(), ".exr" ) )
+	{
+		throw IECore::Exception( "ID manifest file path does not end in \".exr\": " + filePath.generic_string() );
+	}
```
Is this really our business? I think it would be the first place we checked such a thing. It might also upset some upper-case-loving folks.

johnhaddon commented on src/GafferScene/RenderManifest.cpp:
```diff
@@ -0,0 +1,445 @@
+		throw IECore::Exception( "ID manifest file path does not end in \".exr\": " + filePath.generic_string() );
+	}
+
+	if( !filePath.is_absolute() )
+	{
+		throw IECore::Exception( "ID manifest file path is not an absolute file path : " + filePath.generic_string() );
+	}
```
Why is this our business? I don't see why a general `write()` method should refuse to use a relative path.

johnhaddon replied:
Why is this our business? I don't see why a general `write()` method should refuse to use a relative path.

johnhaddon commented on src/GafferScene/RenderManifest.cpp:
```diff
@@ -0,0 +1,445 @@
+		throw IECore::Exception( "ID manifest file path is not an absolute file path : " + filePath.generic_string() );
+	}
+
+	std::filesystem::create_directories( filePath.parent_path() );
```
I think this belongs one level up, in Render and RenderController.

danieldresser-ie replied:
Makes sense. While fixing this, I decided to include 197c04766767c8d9ead5d7a12a4bc8f177701f0f in the RenderManifest PR - it's not technically in the range of commits you requested, but it's a simple change, and the matching change is Render is included, so it makes sense to include it as well.

johnhaddon commented on src/GafferScene/RenderManifest.cpp:
```diff
@@ -0,0 +1,445 @@
+
+	Imf::MultiPartOutputFile exrOutput( filePath.generic_string().c_str(), headers.data(), 1 );
+
+	float image[93] = {
+		1,1,0,0,0,1,1,0,1,1,0,0,0,1,0,0,1,1,1,0,1,1,1,0,0,1,1,0,1,1,1,
+		1,1,1,0,1,0,1,0,1,0,1,0,0,1,0,0,1,1,0,0,1,1,0,0,0,1,0,0,0,1,0,
+		1,0,1,0,1,0,1,0,1,0,1,0,0,1,0,0,1,0,0,0,1,1,1,0,1,1,0,0,0,1,0
+	};
```
Ho ho ho!

johnhaddon reviewed (commented):
I have to confess to rather running out of steam when reviewing a PR of this size. So what I've done is review up to https://github.com/GafferHQ/gaffer/pull/6329/commits/7fa7a4a3826bf8cb52f9fbbad8ef1a18ecb3113e with (hopefully) a good level of attention, and once that is out of the way and merged I will bring my attention to bear on the rest.

I should note a couple of things though, since I ran into problems while testing the whole shebang :

- I think there is still a bug in ImageSelectionTool somewhere where we're not scoping the right context. I'm getting a lot of `Could not open "/renders///id/id.0001.exr" (No such file or directory)` where the gaps in the image file path correspond to context variable substitutions.
- I think we need to add a standard ID output to `startup/gui/outputs.py` to help folks get up and running quickly with this.

Cheers...
John

danieldresser-ie commented on src/GafferSceneModule/RenderManifestBinding.cpp:
```diff
@@ -0,0 +1,79 @@
+namespace
+{
+
+IECore::InternedStringVectorDataPtr pathForIDWrapper( RenderManifest *renderManifest, uint32_t id )
```
Seems reasonable. It does raise a question to me: is it OK that when loading a material cryptomatte, the material names are treated as a path? Seeing the leading slash on the material name in this test of a material cryptomatte looks a bit odd `self.assertEqual( r.pathForID( 1650131652 ), "/shader:ba1b35e0886c8e92a5c5daea71484b84" )`. But it is nice that this API seems to support material cryptomattes well enough that we should be able to use them for picking ... maybe the leading slash isn't really a problem.

danieldresser-ie commented on include/GafferScene/RenderManifest.h:
```diff
@@ -0,0 +1,129 @@
+
+		uint32_t acquireID( const ScenePlug::ScenePath &path );
+
+		uint32_t idForPath( const ScenePlug::ScenePath &path ) const;
```
Good point. I've added a bit of documentation to all the public methods while I'm at it.

danieldresser-ie commented:
Current feedback addressed in #6402, aside from the comments about effectiveWipePlane, which can be addressed just by duplicating that code in ImageSelectionTool.

I haven't addressed the context bug in ImageSelectionTool, or added the standard ID output to startup/gui/outputs.py yet. I guess that will require individual testing with all the renderers we support?

danieldresser-ie commented:
Rebased onto #6402. Removed effectiveWipePlane. Scoped context in ImageSelectionTool everywhere I could find where it is needed.

Added "id" to render pass setups for Arnold ( should I do this for other renderers too? ).

Should I set a default value for StandardOptions.renderManifestFilePath? Something like `${project:rootDirectory}/renders/${script:name}/${renderPass}/idManifest/idManifest.####.exr`? It feels slightly inconsistent, since most options in StandardOptions don't have a default value that differs from the actual default ... actually, there is precedent for having a special default value that helps understand how the option is used: `overscanLeft` has a default value of `0.1`. That probably means I can do the same here?

danieldresser-ie commented:
Oh, and other than that, the outstanding issue I can think of is hotkey assignment. We want to use `Q` for this, so we need to pick something else for the image comparison tool?

johnhaddon commented:
> Added "id" to render pass setups for Arnold ( should I do this for other renderers too? ).

I was hoping we could have a universal setup, like the Beauty one. We already make all the backends support a "standard" ID output for the raytraced viewer, so I think we could achieve the same for a file-based output.

> Should I set a default value for StandardOptions.renderManifestFilePath? Something like ${project:rootDirectory}/renders/${script:name}/${renderPass}/idManifest/idManifest.####.exr?

The place to do that would be in `startup/gui/project.py` or `startup/gui/outputs.py`, with `userDefault` metadata. The nodes themselves shouldn't have any pipeline-specific opinions - that is what config files are for.

I'm in two minds about whether or not the default would be a good thing or not. On the one hand, I want to make this feature as easy to use as possible, but on the other I'm wary of sudden additional overhead that hasn't been explicitly opted in to. Is there any merit in only writing the manifest if a valid ID output has also been specified? So the only step to enabling the workflow is adding the ID output?

> We want to use Q for this, so we need to pick something else for the image comparison tool?

That sounds right to me. @murraystevenson probably has a better sense of what would be a good fit than I do, but for the sake of argument I'll propose "K" for "Kompare" :)

johnhaddon commented:
> Rebased onto https://github.com/GafferHQ/gaffer/pull/6402.

I've merged #6402 to `main` now, so if you rebase on `main` this PR should show only the new commits.

danieldresser-ie commented:
Rebased onto `main`.

> I was hoping we could have a universal setup, like the Beauty one. We already make all the backends support a "standard" ID output for the raytraced viewer, so I think we could achieve the same for a file-based output.

I've tried a naive version of this where I just make the Arnold one generic - several roadblocks immediately appear. Neither Cycles or 3delight appear to write integer channels to exr? At least, setting the data to "uint id" results in a integer cast to float. Neither Cycles or 3delight appear to support "header:" output parameters either? Even the "layerName" parameter to force the final channel name to be "id" isn't working - though that at least has code to handle it in our 3delight backend, so it feels more like a bug than a missing feature.

Anyway, feels like there will probably still be some work to get this functional in other renderers - we'll probably need to discuss the right approach.

> Should I set a default value for StandardOptions.renderManifestFilePath? Something like ${project:rootDirectory}/renders/${script:name}/${renderPass}/idManifest/idManifest.####.exr?

For the moment I've put something in outputs.py as suggested. It's the fairly weak version, where it's not on by default, but if you click it on, you've got a default path filled in by outputs.py. We can continue thinking about whether there's a way to simplify this from requiring 2 separate things to be enabled.

>    We want to use Q for this, so we need to pick something else for the image comparison tool?
>That sounds right to me. @murraystevenson probably has a better sense of what would be a good fit than I do, but for >the sake of argument I'll propose "K" for "Kompare" :)

In the absence of a better idea so far, I've moved image comparison to "K". I don't really like that the wipe hotkey is now so far away - toggling comparison and toggling the wipe are quite related, and they used to be Q and W. Murray didn't really like using a key so far away from the left hand - but we haven't come up with a particular other idea yet.

johnhaddon commented on src/GafferImageUI/ImageGadget.cpp:
```diff
@@ -1012,20 +1691,20 @@ void ImageGadget::renderTiles() const
 
 			glBegin( GL_QUADS );
 
-				glTexCoord2f( uvBound.min.x, uvBound.min.y  );
-				glMultiTexCoord2f( GL_TEXTURE1, validBound.min.x, validBound.min.y  );
+				glTexCoord2f( uvBound.min.x, uvBound.min.y );
+				glMultiTexCoord2f( GL_TEXTURE1, validBound.min.x, validBound.min.y );
 				glVertex2f( validBound.min.x * pixelAspect, validBound.min.y );
 
-				glTexCoord2f( uvBound.min.x, uvBound.max.y  );
-				glMultiTexCoord2f( GL_TEXTURE1, validBound.min.x, validBound.max.y  );
+				glTexCoord2f( uvBound.min.x, uvBound.max.y );
+				glMultiTexCoord2f( GL_TEXTURE1, validBound.min.x, validBound.max.y );
 				glVertex2f( validBound.min.x * pixelAspect, validBound.max.y );
 
-				glTexCoord2f( uvBound.max.x, uvBound.max.y  );
-				glMultiTexCoord2f( GL_TEXTURE1, validBound.max.x, validBound.max.y  );
+				glTexCoord2f( uvBound.max.x, uvBound.max.y );
+				glMultiTexCoord2f( GL_TEXTURE1, validBound.max.x, validBound.max.y );
 				glVertex2f( validBound.max.x * pixelAspect, validBound.max.y );
 
-				glTexCoord2f( uvBound.max.x, uvBound.min.y  );
-				glMultiTexCoord2f( GL_TEXTURE1, validBound.max.x, validBound.min.y  );
+				glTexCoord2f( uvBound.max.x, uvBound.min.y );
+				glMultiTexCoord2f( GL_TEXTURE1, validBound.max.x, validBound.min.y );
```
Especially for commits of this size, it's easier on the reader if unrelated formatting changes are split into a separate commit.

danieldresser-ie replied:
Addressed in [5d14e21](https://github.com/GafferHQ/gaffer/pull/6329/commits/5d14e2131c0bce41f3dacce56d53ed3a8c81de11)

johnhaddon commented on include/GafferImageUI/ImageGadget.h:
```diff
@@ -339,12 +339,11 @@ class GAFFERIMAGEUI_API ImageGadget : public GafferUI::Gadget
 
-		class BufferTexture;
-
+		class GAFFERIMAGEUI_API BufferTexture;
```
I'm baffled by this. We have lots of similar seeming private classes elsewhere, and they don't need the export. Any ideas @ericmehl?

danieldresser-ie replied:
That was my initial thought, and why I didn't add this initially. But I'm pretty sure that in other places where this isn't needed, we only expose raw pointers to the private class? Here, because we expose a unique_ptr to the internal class, I guess other things need to know about the destructor? Does feel a bit ugly to be exporting this, when we want it to be as private as possible.

johnhaddon replied:
We use them with `unique_ptr` too - the only requirement I'm aware of there is that the main class destructor is in the `.cpp` so that the private class destructor is available when it's compiled. I don't get why it doesn't work here. Not the biggest deal, but I would like to understand it if possible - @ericmehl, any ideas?

danieldresser-ie replied:
Turns out that not having visibility declaration in the header is fine. The problem was that the .cpp contained a visibility declaration that had been copied from GafferUI, so listed the wrong module. Fixed here: [5320f77](https://github.com/GafferHQ/gaffer/pull/6329/commits/5320f776de0d82d40b0e717a66580cf4d4a277e2)

johnhaddon commented on src/GafferImageUI/ImageGadget.cpp:
```diff
@@ -75,6 +78,401 @@ using namespace GafferUI;
+void renderUnitSquare( const IECoreGL::Shader::Parameter *pParameter, const IECoreGL::Shader::Parameter *uvParameter = nullptr )
+{
+	static float rectPBufferData[12] = { -1, -1, 0,  -1, 1, 0,  1, -1, 0,  1, 1, 0 };
+	static float rectUvBufferData[12] = { 0, 0,  0, 1,  1, 0,  1, 1 };
```
Should the array length be 8 here?

danieldresser-ie replied:
Fixed here:[5320f77](https://github.com/GafferHQ/gaffer/pull/6329/commits/5320f776de0d82d40b0e717a66580cf4d4a277e2)

And here: [989a141](https://github.com/GafferHQ/gaffer/pull/6329/commits/989a1410f25ac999b2af6859a6756af6c9484e60)

johnhaddon commented on src/GafferImageUI/ImageGadget.cpp:
```diff
@@ -75,6 +78,401 @@ using namespace GafferUI;
+{
+	static float rectPBufferData[12] = { -1, -1, 0,  -1, 1, 0,  1, -1, 0,  1, 1, 0 };
+	static float rectUvBufferData[12] = { 0, 0,  0, 1,  1, 0,  1, 1 };
+	static IECoreGL::BufferPtr g_rectPBuffer = new IECoreGL::Buffer( rectPBufferData, sizeof( float ) * 12 );
```
I'm a bit wary of this because it seems like we'll be destroying the GL resource during the destruction of statics at program exit, when there won't be a GL context around. I see we have the same in ViewportGadget though, so maybe it's OK somehow?

danieldresser-ie replied:
Seems to be working fine ... do you have an alternate way you'd prefer me to do it?

johnhaddon commented on src/GafferSceneUI/ImageSelectionTool.cpp:
```diff
@@ -0,0 +1,782 @@
+
+	sampler.visitPixels(
+		validRect,
+		[&result, &prevValue, &wipeEnabled, &wipePosition, &wipeDirection] ( float value, int x, int y )
```
C++17 doesn't allow lambda to capture structured bindings - GCC will let you do it, but Clang refuses to compile. One workaround seems to be to do an "init capture" instead : `&wipeEnabled = wipeEnabled`.

johnhaddon replied:
C++20 allows it of course, but as we all know, it isn't 2020 yet in these parts.

danieldresser-ie commented on src/GafferImage/Catalogue.cpp:
```diff
@@ -566,13 +566,62 @@ class Catalogue::InternalImage : public ImageNode
+
+						std::string manifestFilename = fileName.filename().generic_string();
+
+						// Remove ".exr"
+						manifestFilename.resize( manifestFilename.size() - 4);
```
When I saw this comment, I assumed I was converting the whole path to string, which would be a bit sloppy. But it's only the filename that I'm operating on ... and the reason I'm operating on the the filename instead of the extension is that the extension should still be ".exr" - the "_manifest" suffix should come before the extension. The main thing that feels confusing to me about this code is that the original file path is called "fileName" instead of "filePath" - maybe it would be clearer if I modified the naming of the existing code? Anyway, I can save one line here by using the `.stem` method of `path` which I wasn't familiar with.

danieldresser-ie replied:
Addressed in [179c1db](https://github.com/GafferHQ/gaffer/pull/6329/commits/179c1db0eba5394bee1e9882368b242bd8073c63)

danieldresser-ie commented on src/GafferImageUI/ImageGadget.cpp:
```diff
@@ -1137,12 +1816,41 @@ void ImageGadget::renderLayer( Layer layer, const GafferUI::Style *style, Render
+	// shader that converts regions to outlines.
+	if( m_idChannel != "" && ( m_selectedIDs.size() || m_highlightID != 0 ) )
+	{
+		{
+			RenderTexture::RenderScope bindScope( m_selectionRenderTexture.get(), true );
+			renderTiles( true );
+		}
```
I think I've figured this out, assuming you're triggering it the same way that I noticed it: when using ImageSelectionTool, and viewing the id channel at the same time, this can be triggered when the image is dirtied. Currently the ImageGadget::Tile::texture() call just takes a bool for whether to convert to floats or ints ... but if the same tile is accessed as a float, and then as an int ... it just pulls the already converted texture, and doesn't notice it's the wrong type. Haven't yet decided how to redesign this.

johnhaddon replied:
Ah, yes, I think that does explain what I've been seeing.

danieldresser-ie commented on src/GafferImage/Catalogue.cpp:
```diff
@@ -566,13 +566,62 @@ class Catalogue::InternalImage : public ImageNode
+
+					IECore::ConstCompoundDataPtr metadata = m_imageCopy->outPlug()->metadata();
+					const StringData *manifestPath = metadata->member<StringData>( "gaffer:renderManifestFilePath" );
+					if( manifestPath && manifestPath->readable().size() )
```
I got most of the way through figuring out the approach I suggested in the meeting on Thursday: if the image is currently connected to a live InteractiveRender, just write the current live manifest straight to the destination path. This just about works ... except that GafferImage::Catalogue shouldn't depend on GafferScene. I guess I'd have to do a bunch of callback registry stuff for calling SceneAlgo::sourceScene, testing if a node is an InteractiveRender and calling InteractiveRender::renderManifest, and holding a shared pointer to a RenderManifest?

Maybe to make the last one easier, RenderManifest should be in Gaffer, not GafferScene? The only thing that makes it specific to GafferScene is using vectors of InternedString to store paths, and that convention is already shared with non-scene specific classes like IECore::PathMatcher?

johnhaddon replied:
It sounds pretty fiddly, even with RenderManifest moved to Gaffer. And RenderManifest really doesn't sound like it belongs in Gaffer.

I wonder what you think to moving Catalogue to GafferScene so we can just do what we want directly? Although technically it can be used without GafferScene, in practice it is always used with InteractiveRender etc. So thematically I think having it in GafferScene is more reasonable than having RenderManifest in Gaffer. It wouldn't be the first ImageNode in GafferScene either - we have Cryptomatte already.

danieldresser-ie commented on python/GafferSceneUI/StandardOptionsUI.py:
```diff
@@ -518,6 +518,20 @@ def __statisticsSummary( plug ) :
+
+		"description",
+		"""
+		Specifies a file to write a sidecar render manifest to. If a
```
Addressed in prereq commit already.

danieldresser-ie commented on python/GafferArnoldTest/ArnoldRenderTest.py:
```diff
@@ -1640,5 +1640,49 @@ def _createPointLight( self ) :
 		return light, light["parameters"]["color"]
 
+	@GafferTest.TestRunner.PerformanceTestMethod( repeat = 1 )
+	def testManySimpleObjectIDs( self ) :
```
Moved to [9914b23](https://github.com/GafferHQ/gaffer/pull/6329/commits/9914b23b38b7d24ef136656132ff406adfd5397f)

danieldresser-ie commented on src/GafferImage/Catalogue.cpp:
```diff
@@ -566,13 +566,62 @@ class Catalogue::InternalImage : public ImageNode
+						{
+							// The most likely cause of this is snapshotting a render in progress, which
+							// won't have written its manifest yet. If we wanted this to work, we would
+							// need to trigger a signal in InternalImage::driverClosed which something
```
This was all reworked in [6edee27](https://github.com/GafferHQ/gaffer/pull/6329/commits/6edee272fd1678c9875e43325d76c28466263e50)

danieldresser-ie commented on startup/gui/outputs.py:
```diff
@@ -76,6 +76,44 @@
+		{
+			"filter" : "closest",
+			"layerName" : "id",
+			"header:gaffer:loadAsUint" : True,
```
Removed now that we're storing ids on disk as bitcast floats.

danieldresser-ie commented on startup/gui/outputs.py:
```diff
@@ -76,6 +76,44 @@
+)
+
+GafferScene.Outputs.registerOutput(
+	"Batch/id",
```
Addressed in prior PR.

danieldresser-ie commented on startup/gui/outputs.py:
```diff
@@ -76,6 +76,44 @@
+			"filter" : "closest",
+			"layerName" : "id",
+			"header:gaffer:loadAsUint" : True,
+			"header:gaffer:integerMapping" : "reinterpret",
```
Removed now that we're storing ids on disk as bitcast floats.

danieldresser-ie commented on startup/gui/outputs.py:
```diff
@@ -76,6 +76,44 @@
+			"filter" : "closest",
+			"layerName" : "id",
+			"header:gaffer:loadAsUint" : True,
+			"header:gaffer:integerMapping" : "reinterpret",
```
Irrelevant now that we're storing ids on disk as bitcast floats.

danieldresser-ie commented on include/GafferImageUI/ImageGadget.h:
```diff
@@ -167,6 +170,10 @@ class GAFFERIMAGEUI_API ImageGadget : public GafferUI::Gadget
 
+		void setSelectedIDs( const std::vector<uint32_t> &ids );
+
+		void setHighlightID( uint32_t id );
```
Addressed in: https://github.com/GafferHQ/gaffer/pull/6329/commits/5320f776de0d82d40b0e717a66580cf4d4a277e2

danieldresser-ie commented:
Addressed comments so far. Added Changelog. Added [25f1005](https://github.com/GafferHQ/gaffer/pull/6329/commits/25f100537079da4469c8f160ce84270d7b9e0278) ( if I recall correctly, the intention was that if we were going to support this behaviour for ImageSelectionTool, we should also support it for SelectionTool, and I think you said you were happy with it while testing with ImageSelectionTool ).

I think the only comment here I haven't addressed is this: "I'm not sure I'm 100% sold on the status bar jumping in size as the path length of the hovered object changes though - would be interesting to see how other folks feel about that." I guess the alternative would be something more like a status bar that is permanently fixed to the bottom of the window? Would it still flick off when you're not hovering anything? I guess this could be something for the meeting tomorrow.

johnhaddon commented on src/GafferSceneUI/SelectionTool.cpp:
```diff
@@ -407,7 +407,12 @@ IECore::RunTimeTypedPtr SelectionTool::dragBegin( GafferUI::Gadget *gadget, cons
-	if( !sg->objectAt( event.line, objectUnderMouse ) )
+	// We treat drags as region selects ( instead of dragging the value of current selection ) if
+	// shift or control is held or the drag starts over the background.
+	if(
+		( event.modifiers & DragDropEvent::Shift ) || ( event.modifiers & DragDropEvent::Control ) ||
+		!sg->objectAt( event.line, objectUnderMouse )
+	)
```
Sorry, I don't think I tested this thoroughly enough previously. We're still handling the click, so the object under the mouse gets selected/deselected before the drag starts, which kindof breaks the experience of the drag being a single gesture. The only way I can see of fixing that is to move the select/deselect to button release?

johnhaddon reviewed (commented):
Thanks for the update Daniel. As discussed yesterday, I've now reviewed everything but the ImageSelectionTool.

Unfortunately I don't think the Catalogue saving code is working as intended. I don't think the manifest from the InteractiveRender node is getting saved when the render stops. My guess is you didn't notice because you're testing with an old on-disk manifest from before InteractiveRender stopped saving them? Or I could be going crazy I suppose. Either way, we could really do with some good test coverage for all the manifest-wrangling shenanigans the Catalogue does - at the moment we have none.

I'm tempted to say that the Catalogue should never do file-copies for manifests. Now we're not writing them to disk from RenderController, the only time it might be useful would be when batch rendering to the Catalogue. But there I think I'm tempted to take the risk on referencing a stale manifest in return for simplifying the wrangling done by the Catalogue, at least for a first version.

Cheers...
John

danieldresser-ie commented on src/GafferScene/Catalogue.cpp:
```diff
@@ -642,8 +761,14 @@ class Catalogue::InternalImage : public ImageNode
 				GafferImage::ImageWriterPtr m_writer;
 
+
+				std::shared_ptr<const GafferScene::RenderManifest> m_liveRenderManifest;
+				std::filesystem::path m_manifestSource;
```
m_manifestSource is gone now that Catalogue is able to deal solely with in-memory manifests.

danieldresser-ie commented on src/GafferScene/Catalogue.cpp:
```diff
@@ -560,19 +601,97 @@ class Catalogue::InternalImage : public ImageNode
+						}
+						else
+						{
+							// If there isn't a live manifest, that should mean that the render is
+							// finished, and there should be a manifest on disk to copy.
+							try
+							{
+								std::filesystem::copy( m_manifestSource, m_manifestDest );
+							}
+							catch( std::filesystem::filesystem_error &e )
+							{
+								// This failure should now be quite unlikely ... it might be possible to
+								// trigger via a race condition where you snapshot an image at the same
+								// instant it finishes rendering. Or by manually deleting the on disk
+								// manifest before snapshotting.
+								IECore::msg( IECore::Msg::Error, "Saving Catalogue image manifest", e.what() );
+							}
```
The current status now is that it is possible to have header:gaffer:renderManifestFilePath metadata added to interactive outputs if you set the option, but this will trigger a warning about setting the renderManifestFilePath option during interactive rendering. Does this make it acceptable? In practice, users should never see this metadata, since the Catalogue overwrites it.

danieldresser-ie commented on src/GafferScene/Catalogue.cpp:
```diff
@@ -362,9 +387,25 @@ class Catalogue::InternalImage : public ImageNode
+				interactiveRenderNode = IECore::runTimeCast<const GafferScene::InteractiveRender>( scenePlug->node() );
+				if( interactiveRenderNode )
+				{
+					liveRenderManifest = interactiveRenderNode->renderManifest();
```
Addressed by adding m_lastRenderManifest.

danieldresser-ie commented on src/GafferScene/Catalogue.cpp:
```diff
@@ -560,19 +601,97 @@ class Catalogue::InternalImage : public ImageNode
+
+					IECore::ConstCompoundDataPtr metadata = m_imageCopy->outPlug()->metadata();
+					const StringData *manifestPath = metadata->member<StringData>( "gaffer:renderManifestFilePath" );
+					if( manifestPath && manifestPath->readable().size() )
```
This has been addressed. ( Whether a live render manifest is found is driven by renderOptions.hasIDOutput in RenderController ).

danieldresser-ie commented:
I think I've addressed the first stage of your feedback here, and our online conversations, in this new PR: https://github.com/GafferHQ/gaffer/pull/6329

Next up will be to rebase this onto that, and then address the ImageGadget/ImageSelectionTool specific stuff.

danieldresser-ie commented:
Rebased now that #6455 is merged.

On the topic of switching selection to mouse up: tackling this in SelectionTool and ImageSelectionTool sounds like a good followup. It should be pretty straightforward coding wise if we can nail down the right behaviour. From the meeting this morning, I think the outstanding question is what should happen when doing a normal left-click drag from an unselected object.

A drag from a selected object does nothing to selection, and drags the paths from all selected objects.

Currently, dragging from an unselected object selects that object on mouse down and then drags its path.

If we are no longer selecting on mouse down, then that doesn't quite make sense any more.

My proposal was to drag the path to the clicked on object, without selecting it, but John pointed out that this doesn't give much feedback as to which path you're dragging ( you could misclick outside the selected object and think you're dragging the selection, but you actually got a different object ).

I think John's counter-proposal was to select the object under the mouse not on mouse down, but when the mouse starts moving to initiate a drag? That sounds weird to me, but I could try implementing it.

Asking Ivan about it, he says he always selects things before dragging the path from them, and suggested not worrying too much about the case of dragging from an unselected object. I guess based on that logic, we could treat a regular left click drag starting from an unselected object as a region select? That would give clear feedback ... it would be very obvious when you see the region select UI appear that you're not dragging a scene path.

johnhaddon commented:
> Anyways, as for this PR, I've rebased it again onto current main, it should be ready for review.

Thanks! I haven't taken a particularly in-depth look at the code, but I've done a fair bit of testing and all seems to be working pretty nicely, so I'm going to merge. I do have some observations on the usability of the status bar, but since work will be ongoing to add instance picking, we can address those as we go :

- When there's an error state, the status only shows when the mouse is over the image, and when the id is non-zero. This makes it hard to read the status, and impossible to copy/paste the message (which would be useful for checking the filename). We should just show the error unconditionally, no matter where the mouse is.
- The error message is too verbose (see below). I propose we just get straight to the point with `Manifest "{fileName}" does not exist`.
- There seems to be no messaging for the case where the ID channel doesn't exist.
- It's a little bit confusing that images have the manifest path metadata in them, even when the Render node has refused to write the manifest due to lack of an ID AOV.

![image](https://github.com/user-attachments/assets/b6d65a0b-c409-4243-8aa9-75ba240a1a32)

Cheers...
John
